finagle-benchmark/src/main/scripts/build_graphs.R

```diff
@@ -0,0 +1,91 @@
+This is not really an R script, it is a literate transcript of the R
+functions I wrote to generate a graph combined with some commentary on
+the data generated. How I use this script is to copy and paste
+snippets into my R workspace.
+
+# returns a data.frame with response times from different load balancers, sampled at the given rate.
+buildData <- function(rr, p2c, ewma, samples=1000) {
+    rr <- read.table(rr)
+    p2c <- read.table(p2c)
+    ewma <- read.table(ewma)
+    data = data.frame(seq(1, samples),
+         sample(rr$V1, samples),
+         sample(p2c$V1, samples),
+         sample(ewma$V1, samples))
+
+    names(data) <- c("id", "rr", "p2c", "ewma")
+
+    return(data)
+}
+
+allGood <- buildData(rr="all_good_1000_rr.txt", p2c="all_good_1000_p2c.txt", ewma="all_good_1000_ewma.txt")
+
+slowMiddle <- buildData(rr="slow_middle_1000_rr.txt", p2c="slow_middle_1000_p2c.txt", ewma="slow_middle_1000_ewma.txt")
+
+slowStart <- buildData(rr="slow_start_1000_rr.txt", p2c="slow_start_1000_p2c.txt", ewma="slow_start_1000_ewma.txt")
+
+# Overlapping ewma vs rr histograms
+> nines = c(0.5, 0.9, 0.95, 0.96, 0.97, 0.98, 0.99, 0.999, 0.9999, 0.99999)
+> quantile(slowMiddle$p2c, nines)
+     50%      90%      95%      96%      97%      98%      99%    99.9%   99.99%  99.999%
+ 174.000  184.000  189.000  191.000  193.000  197.000  205.000 1723.004 1825.201 1907.291
+> quantile(slowMiddle$ewma, nines)
+     50%      90%      95%      96%      97%      98%      99%    99.9%   99.99%  99.999%
+ 174.000  183.000  188.000  189.000  191.000  194.000  199.000  211.000 1780.103 1819.110
+> quantile(slowMiddle$rr, nines)
+    50%     90%     95%     96%     97%     98%     99%   99.9%  99.99% 99.999%
+ 175.00  187.00  203.00 1664.00 1673.00 1683.00 1714.00 1862.00 1977.70 1994.56
+
+### How to think about the latency graphic where Round Robin jumps up
+to 2 seconds after 95%
+
+What does it mean for our service? If your timeout was 1 second, your
+success rate dropped to 95% during this period of time if you used
+round robin. Your success rate only drops to 99.9% if you used EWMA.
+99% if you used queue-depth as your metric.
+
+We setup a load balancing simulation with repsonse latencies we
+gathered from pinging two datacenters. Ping times were about 175ms
```

Comment:
It's not really relevant if this is between data centers or not. I'm not sure if it was, looking at some notes from steve gury who collected the data, it just says "distribution of ping latency on some remote server". If we want more control over the distribution we can eventually synthesize the latency in Simulation to fit a pre-defined distribution.